app/lib/report_helper.rb

```diff
@@ -32,29 +32,52 @@ def compute_taxon_zscores(report)
     zscore_array
   end
 
-  def taxonomy_details(view_level, report, params)
+  def taxonomy_details(report, params)
     taxon_zscores = compute_taxon_zscores(report)
-    wanted_level = view_level == 'species' ? TaxonCount::TAX_LEVEL_SPECIES : TaxonCount::TAX_LEVEL_GENUS
-    data = taxon_zscores.group_by { |h| h[:tax_level] == wanted_level && h[:hit_type] }
+    genus_level = TaxonCount::TAX_LEVEL_GENUS
+    species_level = TaxonCount::TAX_LEVEL_SPECIES
+    data = taxon_zscores.group_by { |h| h[:tax_level] == species_level && h[:hit_type] }
+    data2 = taxon_zscores.group_by { |h| h[:tax_level] == genus_level && h[:hit_type] }
     nt = data['NT'] || []
     nr = data['NR'] || []
-    htc = highest_tax_counts(nt, nr)
+    nt_genus = data2['NT'] || []
+    nr_genus = data2['NR'] || []
+    genus_nt_nr = nt_genus.concat nr_genus
+
     # filter and sort the nt_scores
+    htc = highest_tax_counts(nt, nr)
     rp = resolve_params(params, htc)
-    nt.keep_if do |h|
+    nt_nr = nt.concat nr
+
+    nt_nr.keep_if do |h|
       (h[:tax_id] >= 0 &&
         h[:zscore] >= rp[:nt_zscore_threshold][:start] &&
         h[:zscore] <= rp[:nt_zscore_threshold][:end] &&
         h[:rpm] >= rp[:nt_rpm_threshold][:start] &&
         h[:rpm] <= rp[:nt_rpm_threshold][:end]
       )
     end
-    sort_report!(nt, params[:sort_by])
-    nr_zscore_by_taxon = nr.group_by { |h| h[:tax_id] }
+    sort_report!(nt_nr, params[:sort_by])
     tax_details = []
-    nt.each do |h|
-      x = nr_zscore_by_taxon[h[:tax_id]]
-      tax_details.push(nt_ele: h, nr_ele: x && x[0])
+    nt_nr.each do |h|
```

Comment:
outside the loop
```
all_tax_ids = nt_nr.map {|h| h[:tax_id]}
tax_ids_str = ""
all_tax_ids.each tax_id do
  tax_ids_str += tax_id.to_s
  tax_ids_str += ","
end
lineage_arr = TaxonLineage.connection.select_all("select tax_id, superkingdom_taxid,  genus_taxid from taxon_lineage where tax_id in ({#tax_ids_str})").to_hash
taxon_name_arr = TaxonLineage.connection.select_all("select tax_id, name from taxon_name where tax_id in ({#tax_ids_str})").to_hash
lineage_info = lineage_arr.group_by {|h| h[:tax_id]}
taxon_name_info = taxon_name_arr.group_by {|h| h[:tax_id]}
```

inside the loop
```
   category_taxid = lineage_info[my_tax_id] && lineage_info[my_tax_id][0]["superkingdom_taxid"]
   genus_taxid = lineage_info[my_tax_id] && lineage_info[my_tax_id][0]["genus_taxid"]
   taxon_name = taxon_name_info[my_tax_id][0]["name"]
```

still not the best way

best to pull this data at the start when you are doing the earlier SQL query using a join

Reply:
Above in the bigger SQL query
```
TaxonCount.connection.select_all("select tax_id, pipeline_output_id, tax_level, count, name, count_type from taxon_counts where pipeline_output_id = #{pipeline_output_id}").to_hash
```
you may pull the additional fields you need with a JOIN, like so:
```
TaxonCount.connection.select_all("select tax_id, pipeline_output_id, tax_level, count, name, count_type, `taxon_lineage`.`genus_tax_id` as `genus_tax_id` from taxon_counts INNER JOIN `taxon_lineage`  ON `taxon_lineage`.`tax_id` = `taxon_counts`.`tax_id` where pipeline_output_id = #{pipeline_output_id}").to_hash
```

Reply:
There is also a RAILS-native way of doing this.